[DDP] Remove train call to module copies

rohan-varma opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #58603 [Reducer] Remove replica size == 1 checks
* **#58595 [DDP] Remove train call to module copies**
* #58594 [Reducer] move comment to the right place
* #58593 Format reducer.cpp, hpp
* #58592 [Reducer] Completely remove VariableIndex

No longer needed since this list is always of size 1.

Differential Revision: [D28548426](https://our.internmc.facebook.com/intern/diff/D28548426/)

rohan-varma committed:
[DDP] Remove train call to module copies

No longer needed since this list is always of size 1.

Differential Revision: [D28548426](https://our.internmc.facebook.com/intern/diff/D28548426/)

[ghstack-poisoned]

zhaojuanmao commented on torch/nn/parallel/distributed.py:
```diff
@@ -892,8 +892,6 @@ def gather(self, outputs, output_device):
     def train(self, mode=True):
         super(DistributedDataParallel, self).train(mode)
-        for module in self._module_copies[1:]:
-            module.train(mode)
```
we still need to keep train(mode) call?

rohan-varma replied:
We have the call to the parent class method on L894, which does `train(mode)` on the module for this process. this call is for the rest of the module copies on this process, but there are none.

rohan-varma committed:
Update on "[DDP] Remove train call to module copies"


No longer needed since this list is always of size 1.

Differential Revision: [D28548426](https://our.internmc.facebook.com/intern/diff/D28548426/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[DDP] Remove train call to module copies"


No longer needed since this list is always of size 1.

Differential Revision: [D28548426](https://our.internmc.facebook.com/intern/diff/D28548426/)

[ghstack-poisoned]
